AUI Toolbar Changes

imciner2 opened the pull request:
This PR makes two sets of changes to the AUI toolbar, which nominally breaks a bit of the API in some places (but a lot of effort was made to preserve API for most of the changes.

The first change is to change the container type used by wxAuiToolBar to be a list of pointers instead of an array of items. This allows wxAuiToolBarItem to be subclassed, and then for those subclasses to be used in the toolbar. This is similar to how a wxMenu operates (and since wxMenu was built that way, I initially thought the toolbar would work that way also). I actually do have a usecase for subclassing a tool, specifically to add multiple kinds of additional data/metadata to a tool (so it is better to subclass it than use a struct in the client data).

This change to a list from an array also allows for quite a bit of modernization of the loops in the AUI toolbar, replacing the older ones using indices with range-based for loops over the items. (the previous array implementation didn't have the proper iterators for range-based for loops for this code).

The second set of changes adds AUI-specific item IDs and actually exports them. The AUI toolbar uses 7 different values for kind, but only 4 were ever exported (since they were also in `wxItemKind`), so the other three were internal-only. This change exports them all in the user-facing API, while giving them the prefix `wxAUI_TB_ITEM_*`.


Overall, I have tried to preserve as much API as I can, but did decide to break the API in the toolbar art class in a non-backwards compatible way (but the fix is simple). This means that I did leave the original overflow item set method as a deprecated method (and marked as such). In order to preserve the API for the AUI-specific item ID change, I modified the toolbar `AddTool` methods to use `ints` instead of `wxitemKind`, allowing either the existing `wxitemKind` values or the new `wxAuiToolBarItemKind` values to be used (and I have kept the numerical values for those two types linked as well so that existing code still works with no changes).

imciner2 committed:
Change container type for wxAuiToolBar's item storage

Change the container type used by wxAuiToolBar to be a list of pointers
instead of an array of items. This allows wxAuiToolBarItem to be
subclassed, and then for those subclasses to be used in the toolbar.

This also allows for modernizing the code to use range-based for loops,
which weren't supported by the previous array implementation.

imciner2 committed:
Introduce wxAuiToolBarItemKind to replace AUI use of wxItemKind

Not all item kinds for wxAuiToolBarItem were in wxItemKind (or exported
to the user-facing API), so introduce separate AUI toolbar item kinds to
fully describe wxAuiToolBarItem kinds.

The API for wxAuiToolBar::AddTool was loosend to be an int to allow using
either the legacy wxItemKind values or the new wxAuiToolBarItemKind
values as an argument.

imciner2 committed:
Use CanBeToggled helper in AUI Toolbar code

imciner2 committed:
Make wxAuiToolBarItem have virtual methods and be inheritable from.

imciner2 commented:
I also realized that `wxAuiToolBar` was still not virtual, so I have modified it to support inheritance by adding a virtual destructor and virtual methods. I have also made the fields protected so that they can be accessed by subclasses.

One thing I am tempted to do is remove the need to make the AUI toolbar a friend of the item and just use the accessor methods instead, allowing subclasses to then modify those as needed. I will note that my want for subclassing comes more from wanting to add more fields to the item, not modify the fields already there.

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -125,7 +141,9 @@ class WXDLLIMPEXP_AUI wxAuiToolBarItem
     }
 
-    void Assign(const wxAuiToolBarItem& c)
+    virtual ~wxAuiToolBarItem() {}
```
A nitpick but this should be
```suggestion
    virtual ~wxAuiToolBarItem() = default;
```

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -150,87 +168,87 @@ class WXDLLIMPEXP_AUI wxAuiToolBarItem
     }
 
-private:
+protected:
```
Do we really need to do this? We already have accessors to all these fields AFAICS, so this doesn't seem necessary.

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -255,7 +273,9 @@ class WXDLLIMPEXP_AUI wxAuiToolBarItem
 #ifndef SWIG
+// wxAuiToolBarItemArray is deprecated but kept for backwards compatibility
 WX_DECLARE_USER_EXPORTED_OBJARRAY(wxAuiToolBarItem, wxAuiToolBarItemArray, WXDLLIMPEXP_AUI);
+WX_DECLARE_EXPORTED_LIST(wxAuiToolBarItem, wxAuiToolBarItemList);
```
Let's avoid using the macro-based containers for the new code, i.e. please just use `std::vector<wxAuiToolBarItem*>` or maybe `std::vector<std::unique_ptr<wxAuiToolBarItem>>` if it should own the items.

imciner2 replied:
So new containers should all be standard library containers now?

Even though the toolbar should own the items, I don't think using the `unique_ptr` is going to work, since we have to pass this object down to the toolbar art provider to draw the dropdown list. A pure vector with the toolbar itself doing the memory management would work.

vadz replied:
> So new containers should all be standard library containers now?

Yes, there is no reason not to use them any more as we require C++11 and this means we can count on them being available.

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -494,17 +514,29 @@ class WXDLLIMPEXP_AUI wxAuiToolBar : public wxControl
                  const wxBitmapBundle& bitmap,
                  const wxString& shortHelpString = wxEmptyString,
-                 wxItemKind kind = wxITEM_NORMAL);
+                 int kind = wxAUI_TB_ITEM_NORMAL)
```
Why `int` and not `wxAuiToolBarItemKind`?

imciner2 replied:
`int` is used to preserve compatibility with code using the old `wxITEM_*` identifiers here. Changing the type of the argument to `wxAuiToolBarItemKind` forces the use of `wxAUI_TB_ITEM_*` constants and causes a compile error when any `wxITEM_*` constants are used, because they can't be implicitly converted between the enums. Both enums can be implicitly converted to `int` though, so changing the `AddTool` functions to take `int` allows it to accept both the new `wxAUI_TB_ITEM_*` constants and the old `wxITEM_*` constants that existing code is using. (and note that `wxAuiToolBarItem` already treats kind as an `int`` in its API and storage, so this is the only change needed to make it work).

vadz replied:
Yes, but this is really a pity as it's not type-safe at all. Could we provide both overloads, taking the old and the new enum? It's annoying, of course, but I think this would be worth doing.

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -560,6 +592,7 @@ class WXDLLIMPEXP_AUI wxAuiToolBar : public wxControl
     bool GetToolFits(int toolId) const;
     wxRect GetToolRect(int toolId) const;
     bool GetToolFitsByIndex(int toolId) const;
+    bool GetToolBarItemFits(wxAuiToolBarItem* item) const;
```
Should it be
```suggestion
    bool GetToolBarItemFits(const wxAuiToolBarItem* item) const;
```
?

vadz reviewed (commented):
Thanks for the PR, but it seems like this would require changing _all_ the existing code using `wxAuiToolBar`, wouldn't it? If so, i.e. unless I'm missing something, it seems like a too high price to pay.

BTW, if I do miss something and the compatibility breakage is somehow not as bad as it looks, I think we should keep using `wxAuiToolBarItemArray` instead of adding another similar-but-different container class.

But right now I'm 80% sure that we shouldn't break all the existing code like this and we could find some other way to achieve what you want (see my reply to wx-dev).

We could (and should) still apply `wxAuiToolBarItemKind` part of the changes however.

imciner2 commented:
> but it seems like this would require changing all the existing code using wxAuiToolBar

No, this won't require changing existing wxAuiToolBar code that uses the public API - as I said in the initial PR description. The only place in `wxAuiToolBar` this internal type (`wxAuiToolBarItemArray`) was exposed to users was in the `SetCustomOverflowItems` function, which took two arguments of type `wxAuiToolBarItemArray` (one for prepend and one for append). I have kept that version of the function (but marked it as deprecated in the docs) and added a new overload that takes the new `wxAuiToolBarItemList` type for both arguments - so old code setting the overflow items using the old type will still work by calling the existing function, and new code using the new type will call the new overload.

The only place the change breaks the public API is in the function `virtual int wxAuiToolBarArt::ShowDropDown(wxWindow* wnd, const wxAuiToolBarItemList& items) = 0;`, where I replaced the previous version using an argument of type `const wxAuiToolBarItemArray& items` with one of type `const wxAuiToolBarItemList& items`. This one is more tricky to preserve the API for, because it is actually called from the wxAuiToolBar code itself, so after this change we expect the `wxAuiToolBarItemList` version. This is IMO not a large change for downstream users to make, since the number doing custom toolbar art with this function is fairly low, and as a quick experiment, I used the new GitHub code search to look for uses of:

* `: public wxAuiToolBarArt` - no uses outside of people including the upstream wx header in their repositor
* `: public wxAuiGenericToolBarArt` - Only use I found is in https://github.com/open-enterprise-solutions/enterprise/blob/1e399ddf028e8d5e6a453ccd8be8d00210bd49e6/src/engine/core/frontend/theme/luna_auitoolbar.h (they have a submodule pinned to 3.2.2 though, so they seem to be able to update at their own pace)
* `: public wxAuiDefaultToolBarArt` - 35 results, only one I found overloading the specific `ShowDropDown` function was https://github.com/yazici/ginkgocadx/blob/d9241fc2919aec369436c6d82f092f00829d7fc0/cadxcore/wx/wxginkgotoolbar.cpp#L89, which actually looks very similar to the version in the base class (so I can't pick out the difference at first glance).
* `: public wxAuiMSWToolBarArt` - No results

---
Note: To see the public API compatibility for old code and the changes here, you can take the AUI sample from the master branch and replace the AUI sample in this PR, and it still builds and works just fine - no changes needed.

---
Where it gets trickier is the change to the internal type, moving it from `wxAuiToolBarItemArray` to `wxAuiToolBarItemList`. This will affect any downstream users who inherit from `wxAuiToolBar` and then directly access the `m_items` object in the toolbar class. I did another GitHub code search for inheritance from the wxAuiToolBar (e.g. searching for `: public wxAuiToolBar`), and out of around 150 results, the vast majority of subclasses just contain functions calling things like `AddTool`, `AddSeparator`, etc. to do initialization of the toolbars. The cases of code directly accessing the `m_items` member variable that is changing type that I found are:

* https://github.com/manyxu/Phoenix3D_2.1/blob/bfb2e3757bf61ac461aeeda9216bf8c8fdf76d99/Phoenix3D/Tools/Nirvana/PX2wxAui.cpp - a getter function and then a loop over all items.
* https://github.com/open-enterprise-solutions/enterprise/blob/1e399ddf028e8d5e6a453ccd8be8d00210bd49e6/src/engine/core/frontend/controls/toolBar.h#L6 - two functions to insert tools/separators at specific points in the toolbar (this access could be fixed by adding insert functions to the `wxAuiToolBar`, which would also bring it into better feature parity with `wxToolBar`, which has those functions).
* https://github.com/yazici/ginkgocadx/blob/d9241fc2919aec369436c6d82f092f00829d7fc0/cadxcore/wx/wxginkgotoolbar.cpp#L315 - direct access to the tool to implement programmatic setting of radio/check tools.
* https://github.com/yazici/ginkgocadx/blob/d9241fc2919aec369436c6d82f092f00829d7fc0/cadxcore/wx/wxginkgotoolbar.cpp#L381 - Custom paint loop that removes dropdown buttons and just draws normal/check/radio items all the same.
* https://github.com/acotty/CodeBlocks_Unofficial_Testing/blob/37cb5bc6fead6e6b3f0a10320a7c6d2070f4ef7f/src/plugins/contrib/wxSmithAui/wxAuiToolBar/wxSmithAuiToolBar.cpp#L40 - A hit test routine to find the tool at a given point
* https://github.com/kirixresearch/strata/blob/85612770e903f855e9b5c9c9d8b4dc051ecc7d38/appmain/linkbar.cpp - Quite a few uses of `m_items` scattered around, but they appear to have made their own member variable called `m_items` that will shadow the base class version, so I don't think these uses will be affected.
---

>  I think we should keep using wxAuiToolBarItemArray instead of adding another similar-but-different container class.

Doing this would completely break compatibility, because then we lose the ability to have the `wxAuiToolBar::SetCustomOverflowItems` function work with pre 3.3.0 code. To preserve that compatibility, we need two different types (keep the old and add the new).

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -345,7 +365,7 @@ class WXDLLIMPEXP_AUI wxAuiToolBarArt
     virtual int ShowDropDown(
                          wxWindow* wnd,
-                         const wxAuiToolBarItemArray& items) = 0;
+                         const wxAuiToolBarItemList& items) = 0;
```
[It looks like](https://cs.github.com/?scopeName=All+repos&scope=&q=wxAuiToolBar+ShowDropDown+-path%3Awx%2Faui%2Fauibar.h+-path%3Awx%2Faui%2Fbarartmsw.h+-path%3Asrc%2Faui%2Fauibar.cpp+-path%3Asrc%2Faui%2Fbarartmsw.cpp) this change is going to break quite some code, as the existing overloads won't be called any longer.

vadz commented:
> > but it seems like this would require changing all the existing code using wxAuiToolBar
> 
> No, this won't require changing existing wxAuiToolBar code that uses the public API

OK, sorry, "all" was clearly a dramatic exaggeration, but it seems like quite a lot of existing code will be affected. Before doing into details, I'd like to reiterate the question asked [in this post](https://groups.google.com/g/wx-dev/c/zv9PhneifbA/m/M0i8WoyKAQAJ): why do we need to make all these changes if it looks like the same goal could be easily achieved simply by using the client data inside the tool?

> The only place the change breaks the public API is in the function `virtual int wxAuiToolBarArt::ShowDropDown(wxWindow* wnd, const wxAuiToolBarItemList& items) = 0;`, where I replaced the previous version using an argument of type `const wxAuiToolBarItemArray& items` with one of type `const wxAuiToolBarItemList& items`. This one is more tricky to preserve the API for, because it is actually called from the wxAuiToolBar code itself,

In principle this is possible but it would require making a temporary copy of `m_items` just to call the deprecated function, so this is going to be ugly and inefficient.

> This is IMO not a large change for downstream users to make, since the number doing custom toolbar art with this function is fairly low, 

I think you might be underestimating this number, see the link in my comment. And not all code is on GitHub, of course, so, typically, if I find more than one or two occurrences on GitHub using some code it's a good indication that non-trivial amount of people use it.

> Where it gets trickier is the change to the internal type, moving it from `wxAuiToolBarItemArray` to `wxAuiToolBarItemList`. This will affect any downstream users who inherit from `wxAuiToolBar` and then directly access the `m_items` object in the toolbar class. I did another GitHub code search for inheritance from the wxAuiToolBar 

[My own search](https://cs.github.com/?scopeName=All+repos&scope=&q=wxAuiToolBar+m_items+-path%3Awx%2Faui%2Fauibar.h+-path%3Awx%2Faui%2Fbarartmsw.h+-path%3Asrc%2Faui%2Fauibar.cpp+-path%3Asrc%2Faui%2Fbarartmsw.cpp) finds ~100 of occurrences which is less than 150 you found, but IMO is still quite sufficient to avoid breaking all of it unless we really have no choice.

And, to round back to the beginning, it looks like we do have the choice of using the client data, so why not do it?

> > I think we should keep using wxAuiToolBarItemArray instead of adding another similar-but-different container class.
> 
> Doing this would completely break compatibility, because then we lose the ability to have the `wxAuiToolBar::SetCustomOverflowItems` function work with pre 3.3.0 code. To preserve that compatibility, we need two different types (keep the old and add the new).

OK, yes, this is a good point. But having 2 similar containers is still ugly.

vadz commented:
To return to this, unless there are some crucial advantage of this compared to using client data, I'm afraid that applying it is not worth the breakage, sorry.

Please reopen if you think this is wrong, but I'd really like to see at least some (and preferably convincing) arguments to justify doing this.
